# Bali MEI practical demonstration

### Creating a virus genome from FASTQ reads

Here we'll go through the steps taken to quality control reads that have been generated by a MiSeq run, and create a consensus genome.

These steps require you to run commands in a UNIX-based operating system, for example a Mac OS or Linux. For this practical I will assume you are using a Mac OS.

I would **strongly** recommend that you familiarise yourself with working on the command-line by following the `Learn the Command Line` course at Codecademy: https://www.codecademy.com/learn/learn-the-command-line

1. Download the necessary files to follow this demonstration by clicking on the `Download ZIP` file at the top right of the main page.

2. Open the `Terminal` app, which gives you access to the command-line environment that we will be working in. in `Finder` it's located under `Applications/Utilities/Terminal.app`.

3. Move to the folder you downloaded in step 1:

  `cd /Downloads/Bali-MEI-demonstration-master/FASTQ-to-FASTA/`

  In here you'll see a folder containing the NGS reads in a compressed FASTQ file. There are 2 files: `FASTQ-files/83_S1_L001_R1_001.fastq.gz` contains the forward reads while `FASTQ-files/83_S1_L001_R1_001.fastq.gz` contains the reverse reads of this paired-end run.

4. Check a read by displaying the first 4 lines of the file:

  `gunzip -c FASTQ-files/83_S1_L001_R1_001.fastq.gz | head -4`

  You should see the following:

  ```
  @M00221:25:000000000-ABVKB:1:1101:17060:1399 1:N:0:1
  AGCGAAAGCAGGGGGTTTTACTTTCTGCTCAGTTTTGCTGTAAGCCAAAACTCCTCTAAAGAAAGATGTATATGAAAGAAGGAAGAGATAGAAAGGCTGAGAACTACCAAGGCTAAAACTTCCCTGATCCTATAACCTCAAACACAAACAG
  +
  CCCCCCCBFFFCGGGGGGGGGGHHGHHHHHHHHGHHHHHHHHHHHHHHHHHHHHHHHHHHHHGHHHHHHHHHHHHHHHHHHHHHHHHHGHHHHHHEGHGHHHHHFHHHHHHFHHHHHHHHHHHHHHHHGHHHHHHHHHHHHHHHHHGHHHE
  ```

  The first line contains the unique read header, the second line is the sequence. The `+` marks the end of the sequence and the beginning of the quality scores for each base in the sequence.

  The first thing we will do is to quality control the reads. This involves trimming the reads from their 3' end until the average quality score of the read is above some value, and removing non-biological sequences such as Illumina adapters and PCR primer sequences. To do this, I have included the jar file for QUASR:

5. Run QUASR's `readsetProcessor.jar` without any arguments, which prints to the screen a list of its options:

  `java -jar programs/QUASR_v7.03/readsetProcessor.jar`

  It should look like the following:

  ![QUASR QC](img/QUASR.png?raw=true)

  We're going to quality control the reads by trimming the 3' of each read until the median quality score of the read is greater than 30.0. If this results in a read shorter than 120 nucleotides, then we're going to throw the read away. As this was a paired-end run, and many genome assemblers require matching forward and reverse reads in each file, if either read fails, both the forward and reverse reads are thrown away.

6. Run QUASR with the following settings:

  `java -jar programs/QUASR_v7.03/readsetProcessor.jar -i FASTQ-files/83_S1_L001_R1_001.fastq.gz -r FASTQ-files/83_S1_L001_R2_001.fastq.gz  -q -l 120 -m 30.0 -o quality-control/Bali-MEI`

  This will create 2 FASTQ files in the `quality-control` folder called `Bali-MEI.qc.f.fq` and `Bali-MEI.qc.r.fq`, containing the forward and reverse quality-controlled reads respectively. You will see that 4.25% of the reads were discarded because they failed quality-control, leaving us with 300993 read pairs.

  Now we want to map these reads against a reference to assemble a genome. To do this, we're going to use a program called SMALT (http://www.sanger.ac.uk/science/tools/smalt-0).

7. SMALT version 0.7.6 has been included in the `programs` folder. To run the script, type:

  `programs/smalt-0.7.6/src/smalt`

  if it gives an error message then you need to re-compile the code by typing:

  ```
  cd programs/SMALT-0.7.6/
  ./configure
  make
  make install
  cd ../..
  ```

  Mapping reads to a reference with SMALT takes 2 steps. 1) Create an index of the reference sequence 2) Map the reads against this index. We will perform these 2 steps:

8. We need to create an index of our reference genome. As we know the sample is an influenza A H3N2 subtype, a relevant H3N2 reference genome has been downloaded and included in the `reference` folder as `reference.H3N2.fasta`. We will first create the hash index of this reference with the following command:

  `smalt index -k 14 -s 8 reference/reference.H3N2 reference/reference.H3N2.fasta`

  This creates the hash index of the genome by breaking down the genome into 14-nucleotide "words" (kmers), each word 8 nucleotides apart. The index files will be created as `reference.H3N2.smi` and `reference.H3N2.sma` in the `reference` folder.
