# Bali MEI practical demonstration

### Creating a virus genome from FASTQ reads

Here we'll go through the steps taken to quality control reads that have been generated by a MiSeq run, and create a consensus genome.

These steps require you to run commands in a UNIX-based operating system, for example a Mac OS or Linux. For this practical I will assume you are using a Mac OS.

I would **strongly** recommend that you familiarise yourself with working on the command-line by following the `Learn the Command Line` course at Codecademy: https://www.codecademy.com/learn/learn-the-command-line

1. Download the necessary files to follow this demonstration by clicking on the `Download ZIP` file at the top right of the main page.

2. Open the `Terminal` app, which gives you access to the command-line environment that we will be working in. in `Finder` it's located under `Applications/Utilities/Terminal.app`.

3. Move to the folder you downloaded in step 1:

  `cd /Downloads/Bali-MEI-demonstration/FASTQ-to-FASTA/`

  In here you'll see a folder containing the NGS reads in a compressed FASTQ file. There are 2 files: `83_S1_L001_R1_001.fastq.gz` contains the forward reads while `83_S1_L001_R1_001.fastq.gz` contains the reverse reads of this paired-end run.

4. Check a read by displaying the first 4 lines of the file:

  `gunzip -c 83_S1_L001_R1_001.fastq.gz | head -4`

  You should see the following:

  ```
  @M00221:25:000000000-ABVKB:1:1101:17060:1399 1:N:0:1
  AGCGAAAGCAGGGGGTTTTACTTTCTGCTCAGTTTTGCTGTAAGCCAAAACTCCTCTAAAGAAAGATGTATATGAAAGAAGGAAGAGATAGAAAGGCTGAGAACTACCAAGGCTAAAACTTCCCTGATCCTATAACCTCAAACACAAACAG
  +
  CCCCCCCBFFFCGGGGGGGGGGHHGHHHHHHHHGHHHHHHHHHHHHHHHHHHHHHHHHHHHHGHHHHHHHHHHHHHHHHHHHHHHHHHGHHHHHHEGHGHHHHHFHHHHHHFHHHHHHHHHHHHHHHHGHHHHHHHHHHHHHHHHHGHHHE
  ```

  The first line contains the unique read header, the second line is the sequence. The `+` marks the end of the sequence and the beginning of the quality scores for each base in the sequence.

  The first thing we will do is to quality control the reads. This involves trimming the reads from their 3' end until the average quality score of the read is above some value, and removing non-biological sequences such as Illumina adapters and PCR primer sequences. To do this, I have included the jar file for QUASR:

5. Run QUASR's `readsetProcessor.jar` without any arguments, which prints to the screen a list of its options:

  `java -jar programs/QUASR_v7.03/readsetProcessor.jar`

  It should look like the following:

  ![QUASR QC](simonjwatson.github.com/Bali-MEI-demonstration/FASTQ-to-FASTA/img/QUASR.png)
